crates/red_knot/build.rs

```diff
@@ -69,5 +76,20 @@ fn main() {
     let zipped_typeshed_location = format!("{out_dir}{TYPESHED_ZIP_LOCATION}");
 
     let zipped_typeshed = File::create(zipped_typeshed_location).unwrap();
-    zip_dir(TYPESHED_SOURCE_DIR, zipped_typeshed).unwrap();
+
+    let typeshed_path_filter = |path: &Path| {
+        assert!(path.is_relative());
+
+        // If there is an extension, it has to be `.pyi`
+        path.extension()
+            .map(|extension| {
+                extension
+                    .to_str()
+                    .expect("Expected typeshed file extensions to be UTF-8!")
+                    == "pyi"
+            })
+            .unwrap_or_else(|| typeshed_path.join(path).is_dir())
```

Comment:
Doesn't the walkdir entry expose whether it is a directory or not? I find it a bit confusing that we are handling dirs in this predicate